engine/src/retrier.rs

```diff
@@ -400,11 +400,17 @@ where
 							let half_max = max_sleep_duration(initial_request_timeout, attempt) / 2;
 							let sleep_duration = half_max + rand::thread_rng().gen_range(Duration::default()..half_max);
 
-							let error_message = format!("Retrier {name}: Error for request `{request_log}` with id `{request_id}`, attempt `{attempt}`: {e}. Delaying for {:?}", sleep_duration);
-							if attempt == 0 && !matches!(retry_limit, RetryLimit::Limit(1)) {
-								tracing::warn!(error_message);
-							} else {
-								tracing::error!(error_message);
+							let attempt_number = attempt.saturating_add(1);
+							match retry_limit {
+								RetryLimit::Limit(max_attempts) if attempt_number >= max_attempts => {
+									tracing::error!("Retrier {name}: Error for request `{request_log}` with id `{request_id}`, Reached maximum attempts `{attempt_number}/{max_attempts}`: {e:#}");
```

Comment:
We don't need to log that the maximum attempt reached here - it's logged below when we actually remove the request.

Reply:
But it was displaying "Delaying for ..." which was deceptive, because it was the last attempt.

Reply:
But it still does get delayed here?

Reply:
```rust
			// Delay the request before the next retry.
							retry_delays.push(Box::pin(
								async move {
									tokio::time::sleep(sleep_duration).await;
									// pass in primary or secondary so we know which client to use.
									(request_id, request_log, attempt, retry_limit, primary_or_secondary)
								}
							));
```

Reply:
Basically there's no need to dupe the conditions between the two branches, that was the reason it was implemented this way - keep it in one place where possible